src/test/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/KubernetesAffinityTest.java

```diff
@@ -0,0 +1,236 @@
+import org.csanchez.jenkins.plugins.kubernetes.affinities.NodeAffinity;
+import org.csanchez.jenkins.plugins.kubernetes.affinities.PodAffinity;
+import org.csanchez.jenkins.plugins.kubernetes.affinities.PodAntiAffinity;
+import org.jenkinsci.plugins.workflow.cps.CpsFlowDefinition;
+import org.jenkinsci.plugins.workflow.job.WorkflowJob;
+import org.jenkinsci.plugins.workflow.job.WorkflowRun;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.TemporaryFolder;
+
+/**
+ * @author Carlos Sanchez
+ */
+public class KubernetesAffinityTest extends AbstractKubernetesPipelineTest {
+
+    private static final Logger LOGGER = Logger.getLogger(KubernetesAffinityTest.class.getName());
+
+    @Rule
+    public TemporaryFolder tmp = new TemporaryFolder();
+
+    //Node & Pod affinities to follow
+
+    @Test
+    public void runInPodWithNodeAffinity() throws Exception {
+        deletePods(cloud.connect(), Collections.emptyMap(), false);
+
+        WorkflowJob p = r.jenkins.createProject(WorkflowJob.class, "pod with node affinity");
+        p.setDefinition(new CpsFlowDefinition(loadPipelineScript("runInPodWithNodeAffinity.groovy"), true));
+        WorkflowRun b = p.scheduleBuild2(0).waitForStart();
+        assertNotNull(b);
+
+        Thread.sleep(1000);
+        List<PodTemplate> templates = cloud.getTemplates();
+
+        List<Affinity> nodeAffinities = new ArrayList<>();
+
+        assertFalse(templates.isEmpty());
+
+        for(PodTemplate podTemplate: templates) {
+            if(podTemplate.getLabel().equals("node-affinities")) {
+                nodeAffinities = podTemplate.getAffinities();
+                break;
+            }
+        }
+
+        assertFalse(nodeAffinities.isEmpty());
+
+        //Affinity should be a node affinity
+        Affinity affinity = nodeAffinities.get(0);
+        assertTrue(affinity instanceof NodeAffinity);
+        io.fabric8.kubernetes.api.model.NodeAffinity nodeAffinity = ((NodeAffinity) affinity).buildAffinity();
+
+        assertFalse(nodeAffinity.getPreferredDuringSchedulingIgnoredDuringExecution().isEmpty());
+        assertFalse(nodeAffinity.getRequiredDuringSchedulingIgnoredDuringExecution().getNodeSelectorTerms().isEmpty());
+
+        r.assertBuildStatusSuccess(r.waitForCompletion(b));
+        r.assertLogContains("Apache Maven 3.3.9", b);
+        assertFalse("There are pods leftover after test execution, see previous logs",
+                deletePods(cloud.connect(), KubernetesCloud.DEFAULT_POD_LABELS, true));
+
+    }
+
+    @Test
+    public void runInPodWithPodAffinity() throws Exception {
+        deletePods(cloud.connect(), Collections.emptyMap(), false);
+
+        WorkflowJob p = r.jenkins.createProject(WorkflowJob.class, "pod with pod affinity");
+        p.setDefinition(new CpsFlowDefinition(loadPipelineScript("runInPodWithPodAffinity.groovy"), true));
+        WorkflowRun b = p.scheduleBuild2(0).waitForStart();
+        assertNotNull(b);
+
+        Thread.sleep(1000);
```

Comment:
why the sleeps? what is the condition that must be met ?

Reply:
I had to add the sleep, because it seemed that the template seems to take time to load. This made the test fail as it would not find it on the list of templates.